File: Julia/math/noble_integer.jl
"""Julia program to Find Noble Integer in an array.
Given an array arr, an integer ele is said to be noble integer if
the number of integers greater than ele is equal to ele.
If such an element is not found Return -1. 
"""

function search_noble_integer(arr, n)
    # Traverse the entire array
    for i in 1:(n-1)
        if (arr[i] == arr[i+1])
            continue
        end
        #  If arr[i] equals the number of elements after arr[i], it is a noble Integer
        if(arr[i] == (n - i))
            return arr[i]
        end
    end
    if(arr[n] == 0)
        return arr[n]
    end
    return -1
end


print("How many numbers are present in the array? ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("Array is Empty!!!")
    exit()
end
arr = Int[]
print("Enter the numbers: ")
arr = [parse(Int, num) for num in split(readline())] 
# Sort the array in ascending order
arr = sort(arr)
res = search_noble_integer(arr, n)
if (res == -1)
    print("A Noble Integer is not present in the array.")
else
    print("A Noble Integer $res is  present in the array.")
end


"""
Time Complexity - O(n * log(n)), where 'n' is the size of the array
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I

How many numbers are present in the array? 5
Enter the numbers: 12 34 3 1 99
A Noble Integer 3 is  present in the array.

SAMPLE II

How many numbers are present in the array? 5 
Enter the numbers: 5 4 3 2 1
A Noble Integer is not present in the array.

"""

File: C/math/Noble_Integer.c
/*An integer x is said to be Noble given an array 
if the number of integers greater than x are equal to x.
If noble integer is not found output is -1.
*/
#include <stdio.h>

int nobleint(int *, int);
int main()
{
	int num, i, k, arr[1000], arr2[1000], m;
	printf("Enter the number of elements:");
	scanf("%d", &num);
	printf("Enter the elements:");
	for (i = 0; i < num; i++)
	{
		scanf("%d", &arr[i]);
	}

	m = nobleint(arr, num);
	printf("%d", m);

	return 0;
}

// To find the noble integer from the given array
int nobleint(int *arr, int num)
{
	int count, i, j, x = 0;
	for (i = 0; i < num; i++)
	{
		count = 0;
		for (j = 0; j < num; j++)
		{
			if (*(arr + i)<*(arr + j))
				count = count + 1;
		}

		if (count == *(arr + i))
		{
			x = 1;
			return (count);
		}
	}

	if (x == 0)
		return (-1);
}

/*
Sample Output

Enter the number of elements:4
Enter the elements:7 3 9 81
3
Enter the number of elements:3
Enter the elements:2 2 2
-1

Complexities
Time Complexity:O(n^2)
Space Complexity:O(1)
*/

